zcash_primitives/src/sapling/builder.rs

```diff
@@ -571,11 +568,39 @@ impl InProgressProofs for Proven {
     type OutputProof = GrothProofBytes;
 }
 
+/// Reports on the progress made towards creating proofs for a bundle.
+pub struct ProverProgress {
+    /// The number of steps completed.
+    cur: u32,
+    /// The total number of steps.
+    end: u32,
+}
+
+impl ProverProgress {
+    fn new(cur: u32, end: u32) -> Self {
+        Self { cur, end }
+    }
+
+    /// Returns the number of steps completed so far while creating proofs.
+    ///
+    /// Note that each step may not be of the same complexity/duration.
+    pub fn cur(&self) -> u32 {
+        self.cur
+    }
+
+    /// Returns the total number of steps before the bundle will have valid proofs.
+    ///
+    /// Note that each step may not be of the same complexity/duration.
+    pub fn end(&self) -> u32 {
+        self.end
+    }
+}
+
 struct CreateProofs<'a, SP: SpendProver, OP: OutputProver, R: RngCore> {
     spend_prover: &'a SP,
     output_prover: &'a OP,
     rng: R,
-    progress_notifier: Option<&'a Sender<Progress>>,
+    progress_notifier: Option<&'a Sender<ProverProgress>>,
```

Comment:
Instead of taking a concrete `Sender` struct, could this take a type bounded on a trait that can be implemented for `Sender`? That way it can be implemented as a noop, and if `Sender` is unavailable due to feature flagging or something of the sort, the type constraints can still be satisfied.

Reply:
For example, someone might want to use an implementation of this capability that simply aggregated progress information in memory, rather than sending it over a channel.